GoTo-locator: also add coordinate/crs combi's in result list from crs's from layers (so besides wgs84 and project crs)

rduivenvoorde opened the pull request:
## Description

The rationale behind this PR is that I often work with people who use a EPSG;3857 project with OSM reference map, BUT some national crs (EPSG:28992) details layers. And during (excersises for) emergency responses, they are often offered EPSG:28992 coordinates then to quickly be able to locate something.
One can off course use the perfect LatLon plugin for that (after some fiddling with the preferences to use x,y instead of default y,x... etc), but I thought: why not be able in the locator bar?

My first though: "it would be cool if one could add some crs-indication to the coordinates, so something like: 28992:150000,450000 or so: go to EPSG:28992 X:150000 Y:450000. But my users often are not very EPSG-code aware, and I do not want to make the use of the Locator more difficult.

Then I though: "OK, there is already often EPSG:28992 data, what about not only showing current Project CRS and Wgs84 coordinates, but also crs's which are available in the layer list (and probably of interest for the user...).

So when you have a EPGS:3857 map, and you would paste the valid EPSG:28992 coordinates 108542,472761 into the locator you would be shown:

![image](https://user-images.githubusercontent.com/731673/141683161-43e1bd63-b11f-4363-8a80-282d4996ef65.png)

With this pr it would be:

![Screenshot-20211114142530-778x702](https://user-images.githubusercontent.com/731673/141683202-f0ec5c71-9877-4bfe-a42b-df5bd1ee7e4d.png)

You can also configure one extra crs to test/use via a small configuration dialog (in configure dialog of locators):

![Screenshot-20211214181234-788x197](https://user-images.githubusercontent.com/731673/146047725-cbd4b38a-6a59-4e0e-800f-5822ac3da87a.png)

Note that crs for which the coordinates would not make sense are skipped: 108542,472761 can NOT be valid lat lon coordinates.
I even made this more strict then the old implementation: eg EPSG:28992 has bounds around the Netherlands, with a minimum possible x of say -10000 and a minimal possibly y of say 30000, so when pasting a coordinate like 2,3 it will NOT be shown as possible EPSG:28992 coordinate ( testing a coordinate pair against the bounds of a crs ).

Started of with adding some comments, as it was not clear to me what the scales were for, untill I found out that you can possibly also use osm or google map url's as location... :-) So I added some comments.

I did have some issues creating the logic, so if people think there is a clearer/better way, please let me know.

There are also some caveats in the use of this PR, for example: using the right-mouse click in the map and copying a WGS84 coordinate IN the netherlands, and pasting it in the goto locator, you will NOT be show a EPSG:28992 coordinate, because: the valid lat lon coordinate in NL (say 5.501,52.217) is NOT a valid coordinate in the EPSG_28992 crs... a little counter intuitive maybe?

I also hit the warning: "Used a ballpark transform from EPSG:4326 to EPSG:28992" which you off course do not want to see if you use a simple locator, that is why the logic also became a little complexer...

Let me know what you think, if you think this is troubling the code, I'm ok to thing about alternatives:
- maybe use of 28992:x,y ?
- I event thought about ALSO adding a z or scale value (so you can easily fly to a place AND set the scale) but is was indecisive about using z or scale (and do not want to spoil stuff when working with real 3D data)
- totally forget about this PR (maybe only the comments commit?)

Another thought: as I am testing (potential) xy candidates against the bounds of the crs's: why not also do this for 'Copy coordinate' from mapcanvas? Now it is possible (having a world map (EPSG:4326) loaded, with some national data (EPSG:28992) to try to copy an epsg:28992 coordinate in australia, also raising the ball park warning. While we could also maybe check against the bounds of the crs's before showing them in the context menu?

(sorry for this long text....)

rduivenvoorde committed:
Add some comments in code

rduivenvoorde committed:
Also show coordinate/crs combi's from crs of other layers

Instead of only show possible coordinate combinations of current
MapCanvas crs or Wgs84 crs, this patch also creates a list of all
crs's from all layers and checks IF a coordinate pair would be
possibly a valid coordinate in those crs's

Checking is done by checkin IF a coordinate pair is within the
bounds of the crs, it not, then the resultline is not shown in
the locator results.

rduivenvoorde committed:
Fix Tests

One tests was breaking because we do not have a separate 'Map' and
'crs' results anymore. There is also a little different wording (no 'Map')

The other test was actually invalid in my view because it would result
in 1,234.56 789.012 as WGS84 coordinates

rduivenvoorde commented:
Not sure if my fixing of tests is OK?

In the old way, it returned a result line for (in the Test case) a Map-crs with no value (because not set yet?).
There was also a result with coordinates of "1,234.56 789.012" as being valid ??? coordinates: 1,234.56 789.012. 
I now consider the crs of the layers AND EPSG:4326 so 1,234.56 789.012 is not a valid EPSG:4326 coordinate anymore.

nyalldawson commented on src/app/locator/qgsgotolocatorfilter.cpp:
```diff
@@ -87,51 +88,114 @@ void QgsGotoLocatorFilter::fetchResults( const QString &string, const QgsLocator
-    const bool withinWgs84 = wgs84Crs.bounds().contains( point );
-    if ( !posIsDms && currentCrs != wgs84Crs )
+    // collect all crs's from all layers (plus MapCanvas crs and wgs84)
+    QList<QString> crsList;  // TODO make this a list
+    crsList.append( wgs84Crs.authid() );
```
we should avoid storing QgsCoordinateReferenceSystem objects as strings -- just store them in a QList< QgsCoordinateReferenceSystem > instead.

rduivenvoorde replied:
done in next commit

uclaros replied:
You could also use a QSet instead of a QList and skip checking if CRSs should be added or not

nyalldawson commented on src/app/locator/qgsgotolocatorfilter.cpp:
```diff
@@ -87,51 +88,114 @@ void QgsGotoLocatorFilter::fetchResults( const QString &string, const QgsLocator
-    const bool withinWgs84 = wgs84Crs.bounds().contains( point );
-    if ( !posIsDms && currentCrs != wgs84Crs )
+    // collect all crs's from all layers (plus MapCanvas crs and wgs84)
+    QList<QString> crsList;  // TODO make this a list
+    crsList.append( wgs84Crs.authid() );
```
```suggestion
    QList<QgsCoordinateReferenceSystem> crsList;
    crsList.append( wgs84Crs );
```

nyalldawson commented on src/app/locator/qgsgotolocatorfilter.cpp:
```diff
@@ -87,51 +88,114 @@ void QgsGotoLocatorFilter::fetchResults( const QString &string, const QgsLocator
+    QList<QString> crsList;  // TODO make this a list
+    crsList.append( wgs84Crs.authid() );
+    if ( currentCrs != wgs84Crs )
+      crsList.append( currentCrs.authid() );
```
```suggestion
      crsList.append( currentCrs );
```

nyalldawson commented on src/app/locator/qgsgotolocatorfilter.cpp:
```diff
@@ -87,51 +88,114 @@ void QgsGotoLocatorFilter::fetchResults( const QString &string, const QgsLocator
+      crsList.append( currentCrs.authid() );
+    for ( const QgsMapLayer *layer : QgsProject::instance()->mapLayers() )
+    {
+      if ( !crsList.contains( layer->crs().authid() ) )
```
```suggestion
      if ( !crsList.contains( layer->crs() ) )
```

nyalldawson commented on src/app/locator/qgsgotolocatorfilter.cpp:
```diff
@@ -87,51 +88,114 @@ void QgsGotoLocatorFilter::fetchResults( const QString &string, const QgsLocator
+    {
+      if ( !crsList.contains( layer->crs().authid() ) )
+      {
+        crsList.append( layer->crs().authid() );
```
```suggestion
        crsList.append( layer->crs() );
```

nyalldawson commented on src/app/locator/qgsgotolocatorfilter.cpp:
```diff
@@ -87,51 +88,114 @@ void QgsGotoLocatorFilter::fetchResults( const QString &string, const QgsLocator
+      }
+    }
+    // Now go over all crs's and check if the coordinate COULD be use in that crs
+    for ( const QString &authid : crsList )
     {
+
+      QgsCoordinateReferenceSystem crs = QgsCoordinateReferenceSystem( authid );
```
```suggestion
    for ( const QgsCoordinateReferenceSystem &crs : std::as_const( crsList ) )
    {
```

nyalldawson commented on src/app/locator/qgsgotolocatorfilter.cpp:
```diff
@@ -87,51 +88,114 @@ void QgsGotoLocatorFilter::fetchResults( const QString &string, const QgsLocator
-    if ( withinWgs84 )
-    {
-      if ( currentCrs != wgs84Crs )
+      const bool withinWgs84 = wgs84Crs.bounds().contains( point );
```
This should just be checking the current crs from the list:
```suggestion
      const bool coordinateIsWithinCrs = crs.bounds().contains( point );
```

nyalldawson replied:
ie there shouldn't be a special code path for wgs84

rduivenvoorde replied:
Ok, only reason I see there was (and I copied that) is that the output string says:
```"Go to %1° %2° (%3)"```
So there are little 'degree' symbols...

But I can try to remove that codepath and for example add the 'degree' symbol only if geographical... or I just forget about them anyway....

nyalldawson commented on src/app/locator/qgsgotolocatorfilter.cpp:
```diff
@@ -87,51 +88,114 @@ void QgsGotoLocatorFilter::fetchResults( const QString &string, const QgsLocator
+      }
+
+      // non wgs84 project, create a transform to be able to check point against bounds
+      const QgsCoordinateTransform transform2wgs84( crs, wgs84Crs, QgsProject::instance()->transformContext() );
```
```suggestion
      const QgsCoordinateTransform transform2wgs84( crs, wgs84Crs, QgsProject::instance()->transformContext() );
      transform2wgs84.setBallparkTransformsAreAppropriate( true );
```

rduivenvoorde replied:
Ah, thanks was not aware of this, so no more warnings?
(would be cool to also have something like that for that pesky "no crs defined, will use Lat Lon/WGS84 for layer" (or something like that) ... :-)

nyalldawson commented on src/app/locator/qgsgotolocatorfilter.cpp:
```diff
@@ -87,51 +88,114 @@ void QgsGotoLocatorFilter::fetchResults( const QString &string, const QgsLocator
-    {
-      if ( currentCrs != wgs84Crs )
+      const bool withinWgs84 = wgs84Crs.bounds().contains( point );
+      if ( crs == wgs84Crs && crs == currentCrs && withinWgs84 )
```
Can you re-work this loop to avoid the special handling of wkgs84?

rduivenvoorde replied:
Will  do this/my evening.

rduivenvoorde replied:
yes done in next commit

rduivenvoorde commented:
@nyalldawson Thanks for taking time to review! But in general are you ok with this? Or do you think it is only making stuff complexer (/me like simplicity in general). I mean it IS helpfull to me personally, but I do not want to clutter code.

nyalldawson commented:
>  But in general are you ok with this

Yes, I think it's a great idea!

rduivenvoorde committed:
Handling reviewers comments

rduivenvoorde commented:
Reworked and added all the review comments.

Two things (of which the first one I would like to add in this PR... but please let me know what you think):

- So IF crs coordinates are latlon coordinates, the display string shows degree symbol: 2.344° 48.877°
BUT I see a lot of average users questioning: what is lat, what is lon, is it x,y (when showing degrees),
so I wondered if I should add "lat:" "lon:" to the output
AND if we should maybe (only for 4326 actually SHOW LatLon (instead of x,y/lon,lat what we do now)
According to: https://gis.stackexchange.com/questions/293897/is-srid-4326-lon-lat-or-lat-lon
EPSG:4326 should be LatLon (in that order, instead of LonLat as below):

![Screenshot-20211119162149-423x86](https://user-images.githubusercontent.com/731673/142650716-8bd3382d-2c3c-43f6-aa2c-b271d43ade89.png)

- Second: instead of testing the layer/map crs's, I wondered if I would add a 'configure' item to this locator, to be able (via the 'normal' crs selector), to actually add (some?) crs's.
In my case I would always add EPSG:28992 to it, so I can just use QGIS to fly to a national coordinate.

Opinions?

uclaros commented:
That's a nice and handy feature. Active layer crs should get a higher score than other layers' crss when the coordinates are valid for more than one crs.



> instead of testing the layer/map crs's, I wondered if I would add a 'configure' item to this locator, to be able (via the 'normal' crs selector), to actually add (some?) crs's.


That would also be handy. Such customization can really boost some workflows!

rduivenvoorde commented:
@uclaros so we have (I put it in order, feel free to change)
1) Map crs == 0.9
2) chosen crs (todo: make possible in configs) 0.85
3) active layer 0.80
4) visible layer 0.75
5) other layer crs 0.70 
Something like that?

Any opinion about how to show wgs84 degrees? LatLon-order, or LonLat-order? (see above), Or adding Lat: and Lon: to the coords?

rduivenvoorde commented:
Ok, fiddling with these custom small configuration dialogs:

![Screenshot-20211121195057-468x138](https://user-images.githubusercontent.com/731673/142775140-1a9db5dc-9c86-4406-8945-5ecfcf55098e.png)

Not sure if this is the right use for this QgsProjectionSelectionWidget...

Clicking on the small globe button shows the CRS selector dialog, 

![Screenshot-20211121195411-625x669](https://user-images.githubusercontent.com/731673/142775229-76681ac1-29d8-4086-b718-8eae14fda8ba.png)

and clicking on a CRS then add's it to the 'Recently Used CRSs', which in turn then show up in the QgsProjectionSelectionWidget-dtopdown:

![Screenshot-20211121195617-448x293](https://user-images.githubusercontent.com/731673/142775261-7a1f028a-42d0-45ce-bca6-0910bae20d36.png)

But it looks like this is not meant to be used to create a small list of crs's?
For example: how to remove one from the list?
And how did I end up with these recent ones anyway :-)

rduivenvoorde committed:
Make it possible to select 1 extra crs

Besides the default ones: EPSG:4326, the project crs and available
layers crs's.
The 'extra_crs' is remembered via QgsSettings in your profile

rduivenvoorde commented:
So there is a little configuration dialog now:

![Screenshot-20211214181215-788x210](https://user-images.githubusercontent.com/731673/146047712-dd9afec5-4c5e-4575-9688-18eb3a774dc5.png)

For example selecting EPSG:4258

![Screenshot-20211214181234-788x197](https://user-images.githubusercontent.com/731673/146047725-cbd4b38a-6a59-4e0e-800f-5822ac3da87a.png)

Will be saved into the settings, and then will be show (besides the defaults: epsg:4326, project- and layer-crs's) IF the coordinates is within the valid extent of that extra crs

![Screenshot-20211214181326-629x386](https://user-images.githubusercontent.com/731673/146047984-a4531446-e20f-443b-8f4e-fb422e1562c4.png)

3nids commented on src/app/locator/qgsgotolocatorfilter.cpp:
```diff
@@ -40,15 +42,23 @@ void QgsGotoLocatorFilter::fetchResults( const QString &string, const QgsLocator
+
+  QgsSettings settings;
+  QString extraCrsAuthId = extraCrs.authid();
+  if ( settings.value( QStringLiteral( "locator_filters/go_to_coordinate/extra_crs" ), QStringLiteral( "EPSG:4326" ), QgsSettings::App ) != QStringLiteral( "EPSG:4326" ) )
```
can you declare the setting according to the new guide lines?
https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html#settings

3nids commented on src/app/locator/qgsgotolocatorfilter.cpp:
```diff
@@ -87,51 +99,109 @@ void QgsGotoLocatorFilter::fetchResults( const QString &string, const QgsLocator
+    if ( currentCrs != wgs84Crs )
+      crsList.append( currentCrs );
+    // append optional extra crs (added via config)
+    if ( extraCrs != wgs84Crs )
```
```suggestion
    if ( !crsList.contains( extraCrs ) )
```
to avoid duplicated entries

3nids commented on src/app/locator/qgsgotolocatorfilter.cpp:
```diff
@@ -87,51 +99,109 @@ void QgsGotoLocatorFilter::fetchResults( const QString &string, const QgsLocator
+    if ( extraCrs != wgs84Crs )
+      crsList.append( extraCrs );
+    // append crs's used by mapLayers
+    for ( const QgsMapLayer *layer : QgsProject::instance()->mapLayers() )
```
```suggestion
     const auto constMapLayers = QgsProject::instance()->mapLayers(); 
    for ( const QgsMapLayer *layer : constMapLayers )
```

3nids reviewed (commented):
it would be nice to complete the test case with what you have imlpemented with layers from the layer tree and playing with the different scenarios.

rduivenvoorde commented:
@3nids thanks for looking into this. I'll handle your comments.

About adding tests: I have to find out how to programmatically change the config, but will try.

I see I now have to rebase first and fix conflicts... hope I can make that work :-(

Thanks for (hopefully) make this work...

rduivenvoorde commented:
Argh, really not know how to do this easiest: should I first rebase and then make it work?
Or should I try to mix @nirvn edits into mine....
Anybody an hint?
